mirrord/agent/src/steal/connections/filtered.rs

```diff
@@ -0,0 +1,1524 @@
+        };
+
+        let parts = upgraded
+            .http_client_io
+            .downcast::<TokioIo<T>>()
+            .expect("IO type is known");
+        let mut http_client_io = parts.io.into_inner();
+        let http_client_read_buf = parts.read_buf;
+
+        match upgraded.http_server_io {
+            // Connection was upgraded and some HTTP filter matched the upgrade request.
+            UpgradedServerSide::MatchedClient(client_id) => {
+                tracing::trace!(
+                    connection_id = self.connection_id,
+                    client_id,
+                    "HTTP connection upgraded for client",
+                );
+
+                for (data, _) in queued_raw_data
+                    .into_iter()
+                    .filter(|(_, id)| *id == client_id)
+                {
+                    http_client_io.write_all(&data).await?;
+                }
+
+                for (id, subscribed) in self.subscribed.iter_mut() {
+                    if *subscribed && *id != client_id {
+                        tx.send(ConnectionMessageOut::Closed {
+                            client_id: *id,
+                            connection_id: self.connection_id,
+                        })
+                        .await?;
+
+                        *subscribed = false;
+                    }
+                }
+
+                if self.subscribed.remove(&client_id) != Some(true) {
+                    return Ok(());
+                }
+
+                if !http_client_read_buf.is_empty() {
+                    tx.send(ConnectionMessageOut::Raw {
+                        client_id,
+                        connection_id: self.connection_id,
+                        data: http_client_read_buf.into(),
+                    })
+                    .await?;
+                }
+
+                UnfilteredStealTask {
+                    connection_id: self.connection_id,
+                    client_id,
+                    stream: http_client_io,
+                }
+                .run(tx, rx)
+                .await
+            }
+
+            // Connection was upgraded and no HTTP filter matched the upgrade request.
+            UpgradedServerSide::OriginalDestination(upgraded) => {
+                tracing::trace!(
+                    connection_id = self.connection_id,
+                    "HTTP connection upgraded for original destination",
+                );
+
+                for (client_id, subscribed) in self.subscribed.iter_mut() {
+                    if *subscribed {
+                        tx.send(ConnectionMessageOut::Closed {
+                            connection_id: self.connection_id,
+                            client_id: *client_id,
+                        })
+                        .await?;
+
+                        *subscribed = false;
+                    }
+                }
+
+                let parts = upgraded
+                    .downcast::<TokioIo<TcpStream>>()
```

Comment:
TcpStream not T correct?

Reply:
Yup, this is the connection with the server at the original destination. Made in line 122